duktape/src/main/jni/JavaMethod.cpp

```diff
@@ -19,119 +19,160 @@
-      // unreachable - duk_error never returns.
-      break;
-  }
-  duk_pop(ctx);
-  return value;
-}
-
-/**
- * Convert the given {@code value} to a duktape type and push it onto the stack (for a function
- * call return).
- * Returns the number of entries pushed to the duktape stack.
- */
-duk_ret_t pushJavaScriptType(JNIEnv* env, duk_context *ctx, JavaMethod::Type type, jvalue value) {
-  switch (type) {
-    case JavaMethod::Void:
-      return 0;
-
-    case JavaMethod::String: {
-      const JString result(env, static_cast<jstring>(value.l));
-      duk_push_string(ctx, result);
-      return 1;
-    }
-
-    default:
-      duk_error(ctx, DUK_ERR_API_ERROR, "unsupported return type: %d", type);
-      // unreachable - duk_error never returns.
-      return DUK_RET_INTERNAL_ERROR;
-  }
-}
+JavaMethod::ArgumentLoader getArgumentLoader(JNIEnv *env, jobject typeObject);
+JavaMethod::MethodBody getMethodBody(JNIEnv* env, jmethodID methodId, jobject returnType);
 
 } // anonymous namespace
 
-JavaMethod::JavaMethod(JNIEnv* env, jobject method)
-    : m_methodId(env->FromReflectedMethod(method)) {
+JavaMethod::JavaMethod(JNIEnv* env, jobject method) {
   jclass methodClass = env->GetObjectClass(method);
 
-  const jmethodID getName = env->GetMethodID(methodClass, "getName", "()Ljava/lang/String;");
-  const JString methodName(env, static_cast<jstring>(env->CallObjectMethod(method, getName)));
-  m_name = methodName;
-
-  const jmethodID getReturnType =
-      env->GetMethodID(methodClass, "getReturnType", "()Ljava/lang/Class;");
-  jobject returnType = env->CallObjectMethod(method, getReturnType);
-  m_returnType = getType(env, returnType);
-
   const jmethodID getParameterTypes =
       env->GetMethodID(methodClass, "getParameterTypes", "()[Ljava/lang/Class;");
   jobjectArray parameterTypes =
       static_cast<jobjectArray>(env->CallObjectMethod(method, getParameterTypes));
   const jsize numArgs = env->GetArrayLength(parameterTypes);
-  m_parameterTypes.reserve(numArgs);
+  m_argumentLoaders.reserve(numArgs);
   for (jsize i = 0; i < numArgs; ++i) {
-    m_parameterTypes.push_back(getType(env, env->GetObjectArrayElement(parameterTypes, i)));
+    m_argumentLoaders.push_back(
+        getArgumentLoader(env, env->GetObjectArrayElement(parameterTypes, i)));
   }
+
+  const jmethodID getReturnType =
+      env->GetMethodID(methodClass, "getReturnType", "()Ljava/lang/Class;");
+  jobject returnType = env->CallObjectMethod(method, getReturnType);
+
+  jmethodID methodId = env->FromReflectedMethod(method);
+  m_methodBody = getMethodBody(env, methodId, returnType);
 }
 
-duk_ret_t JavaMethod::invoke(duk_context *ctx, JNIEnv *env, jobject javaThis) const {
-  if (duk_get_top(ctx) != m_parameterTypes.size()) {
+duk_ret_t JavaMethod::invoke(duk_context* ctx, JNIEnv* env, jobject javaThis) const {
+  if (duk_get_top(ctx) != m_argumentLoaders.size()) {
     duk_error(ctx, DUK_ERR_API_ERROR, "wrong number of arguments");
```

Comment:
`duk_error` causes a duktape JS error object to be thrown, so the stack will unwind automatically. The function never returns, but I'll add a return statement for clarity (like I've done elsewhere).